0015-estimator-interface.md

```diff
@@ -0,0 +1,500 @@
+        bindings_array = BindingsArray(bindings_array)
+    elif isinstance(bindings_array, Mapping):
+        bindings_array = BindingsArray(kwargs=bindings_array)
+
+    return bindings_array
+```
+
+In particular, we propose this kind of Coercion for the types:
+* `EstimatorTask`
+* `BindingsArray`
+* `ObservablesArray`
+
+### TaskResults <a name="TaskResults"></a>
+
+The results from each `EstimatorTask` will be array valued, and each `EstimatorTask` in the same job may have a different shape.
+Consider an `EstimatorTask` with shape `<20, 30>`, where the shape has come from the broadcasting rules discussed elsewhere. 
+This will result in a 20×30 array of real estimates.
+Moreover, we will want to return an array of standard deviations of the same shape.
+This would result in a bundle of broadcastable arrays:
+
+```python
+TaskResult({“evs”: <20, 30>, “stds”: <20, 30>}, metadata)
+```
+
+The reason we are proposing a generic container for the return type instead of, e.g., an `Estimator`-specific container, is because it
+
+1. Provides unified experience across primitives for users. 
+2. code-reuse
+3. Provides a certain certain amount of flexibility for what can be returned without modifying the container object. Here are some examples:
+  1. Suppose that we want to give users the option of additionally returning the covariances between estimates that arise because of the circuit multiplexing, then we could update with the field `{"cov": <20,30,20,30>}`.
+  2. Suppose we want to return some indication of which estimates came from the same physical circuit.
+
+## Migration Path <a name="migration-path"></a>
+
+We need to remain backwards compatible with the existing interface to adhere to the [Qiskit Deprecation Policy](https://qiskit.org/documentation/deprecation_policy.html).
+
+In summary, we propose the following strategy: If the user has provided no `TaskLike`s, proceed with the old API, the old API output, and emit a deprecation warning, or an error if something mandatory like `observables` has been omitted. Otherwise, proceed with the new API, raising if they have tried to use the old arguments in addition to providing tasks.
+
+We now describe the strategy in detail, beginning with a restatement of the current interface, where `Estimator.run` has `circuits` as the only positional argument and accepts `observables` and `parameter_values` as keyword arguments (in addition to the var keyword `**run_options` arguments which isn't affected):
+
+```python
+class Estimator(BasePrimitive):
+
+    # current signature
+    def run(
+            self,
+            circuits: Sequence[QuantumCircuit] | QuantumCircuit,
+            observables: Sequence[BaseOperator | PauliSumOp | str] | BaseOperator | PauliSumOp | str,
+            parameter_values: Sequence[Sequence[float]] | Sequence[float] | float | None = None,
+            **run_options,
+        ):
+        ...
+```
+
+We propose the migration to using tasks in two phases.
```

Comment:
As I brought up during the meeting (but didn't get to explore fully due to time), this section covers the backwards compatibility from the estimator end user perspective. But the other axis to consider here is that we need to maintain compatibility for implementers of the base primitive classes. `BaseEstimator` provides an abstract base class that implementers need to subclass to provide an estimator implementation. To provide backwards compatibility it's a requirement to have the interface changes work in a manner that's compatible with **existing** implementations of `BaseEstimator`. Not all implementations of the estimator are part of qiskit itself, and we need to provide a migration path for all downstream implementations to enable using the new interface. Also any internal changes do not interfere with the operation of existing primitives in otherwise previously valid input. 

We can have previously implementations error on task based input (as that wasn't supported before), but that is essentially versioning the interface implicitly and it becomes an exercise for users to negotiate which estimators support which features. In such cases I feel it's much better to be explicit about the features supported via versioning, So, I think there are really 2 paths forward here, either explicitly versioning the `BaseEstimator` interface (i.e. adding a `BaseEstimatorV2` based on what's described in this RFC) and documenting the differences and the end user migration path for consumers of estimators. Then the user can determine which version of the estimator there using and and adjust their usage. Or alternatively we add a parallel internal dispatch mechanism to `BaseEstimator` for the new task based interface so that if a user submits a task style input that is converted into the format existing implementations will support:

https://github.com/Qiskit/qiskit/blob/fe3bffbc85a4fb57dc62ba04fa7f11bcefb5d75f/qiskit/primitives/base/base_estimator.py#L196-L202

Basically this means adding a new optional method like `def _new_run()` which accepts task based input, in the case `_new_run()` is not defined in an implementation `BaseEstimator` will convert the new style input into the old style (likely via a lot of copying) and pass that to `_run()`. This way if end users submit a task based input to `.run()` it will work correctly (albeit inefficiently) even for existing implementations of `BaseEstimator`.

Reply:
For `_run()`, for reference:
 * Qiskit/qiskit#11057

Reply:
I was just discussing this with @Cryoris, that we could maybe avoid the overall interface versioning by having a `def _run_v2()` (or `def _new_run()`), but https://github.com/Qiskit/qiskit/pull/11057 would block that potential path.